### PR TITLE
postgres_db: add dump and restore support

### PR DESCRIPTION
**Pull request by joshmoore**
Originally filed as https://github.com/ansible/ansible-modules-core/pull/2752 by @autotune. 

---

Ended up just recreating the branch with the specific modified file on there per #2751. The issue template has been re-posted below as well:

Issue Type:

Can you help us out in labelling this by telling us what kind of ticket this this? You can say:
Feature Pull Request

Component Name:
postgresql_db module

Ansible Version: 2.1.0

Ansible Configuration:

Standard config.

Environment:

[autotune@ansibledev ansible-modules-core]$ cat /etc/redhat-release
CentOS Linux release 7.1.1503 (Core)

Summary:

This module adds an import and dump state to the core postgresql_db Module that supports .tar, .gz, .xz, .bz, and a standard .sql file:

```

---
- hosts: local
  tasks:
- name:
  2731_postgresql_db: name=ansitest login_user=postgres state=dump login_host=127.0.0.1 target=~/dev/features/2731/ansitest.sql.xz
```

Steps To Reproduce:

Run the above playbook with a database that exists in postgres called "ansitest" and do a dump. Run the same playbook with "state=import" and change name to an new, empty database, and import.

Expected Results:

Either a new dump created in the form of a compressed file or an imported database.

Actual Results:

See above.

### CODE REVIEW

**Comment by joshmoore** [comment]
Migrated from https://github.com/ansible/ansible-modules-core/pull/4614

File: lib/ansible/modules/database/postgresql/postgresql_db.py
Comment (nerzhul) [inline lib/ansible/modules/database/postgresql/postgresql_db.py]: please fix version_added to 2.3

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/database/postgresql/postgresql_db.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_db.py]: Please add a comment on the `description` to say dump and import where added in 2.3

Also is it import or restore?

File: lib/ansible/modules/database/postgresql/postgresql_db.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_db.py]: `"2.3"`

File: lib/ansible/modules/database/postgresql/postgresql_db.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_db.py]: Is it import or restore?

File: lib/ansible/modules/database/postgresql/postgresql_db.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_db.py]: What are these three lines for?
I don't see any `backup` option, also the formatting is wrong.

File: lib/ansible/modules/database/postgresql/postgresql_db.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_db.py]: I don't see a backup option in the code

File: lib/ansible/modules/database/postgresql/postgresql_db.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_db.py]: You seem to alternate between the terms backup, import and restore. Picking two will make this easier for people to follow.

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision

You seem to alternate between the terms backup, import and restore. Picking two will make this easier for people to follow.

**Comment by gundalow** [comment]
@autotune Could you please give this a look. Thanks in advance.

File: lib/ansible/modules/database/postgresql/postgresql_db.py
Comment (joshmoore) [reply lib/ansible/modules/database/postgresql/postgresql_db.py]: Agreed. I didn't even realize the discrepancy when taking over https://github.com/ansible/ansible-modules-core/pull/2752

File: lib/ansible/modules/database/postgresql/postgresql_db.py
Comment (joshmoore) [reply lib/ansible/modules/database/postgresql/postgresql_db.py]: I'm inclined to think these were originally intended as part of the description. I'll migrate.

**Comment by joshmoore** [comment]
I think the latest push addresses all the comments made. Thanks, all.

**Review by gundalow — APPROVED** [review]
shipit

**Comment by gundalow** [comment]
Looks like there are some pep8 tidiups needed (reported by Shippable)

```
2017-01-31 16:45:00 Run command: pep8 --max-line-length 160 --config /dev/null --ignore E123,E124,E127,E128,E201,E202,E203,E211,E221,E222,E225,E226,E227,E228,E231,E241,E251,E261,E262,E265,E266,E301,E302,E303,E402,E502,E713,E731,W391,W503 lib/ansible/modules/database/postgresql/postgresql_db.py
2017-01-31 16:45:00 ERROR: PEP 8: lib/ansible/modules/database/postgresql/postgresql_db.py:262:13: E121 continuation line under-indented for hanging indent (legacy)
2017-01-31 16:45:00 ERROR: PEP 8: lib/ansible/modules/database/postgresql/postgresql_db.py:308:13: E121 continuation line under-indented for hanging indent (legacy)
2017-01-31 16:45:00 PEP 8: There were 3 different legacy issues found (32 total):
2017-01-31 16:45:00 PEP 8: Count Code Message
2017-01-31 16:45:00 PEP 8:    27 E111 indentation is not a multiple of four
2017-01-31 16:45:00 PEP 8:     2 E129 visually indented line with same indent as next logical line
2017-01-31 16:45:00 PEP 8:     3 E501 line too long (NNN > NNN characters)
2017-01-31 16:45:00 ERROR: PEP 8: There are 2 issues which need to be resolved.
```

**Review by nerzhul — COMMENTED** [review]
please fix the last pep8 reports

**Comment by joshmoore** [comment]
Rebase removed & pep8 fixed.

**Comment by joshmoore** [comment]
ready_for_review

**Comment by joshmoore** [comment]
ready_for_review

**Comment by joshmoore** [comment]
bot_status

**Comment by joshmoore** [comment]
(I'm unclear why `needs_revision` was added)

**Comment by joshmoore** [comment]
@gundalow @nerzhul, sorry for the ping, but could either of you clarify? Cheers.

**Comment by nerzhul** [comment]
@joshmoore CI is happy it's good for me

**Comment by joshmoore** [comment]
@gundalow: any thoughts?

**Comment by gundalow** [comment]
Maybe add another ready_for_review

After that @nerzhul needs to give change his review

**Comment by gundalow** [comment]
Also this is a feature request, We've hit feature freeze on 2.3, so this will have to wait till we've hit RC1 (Friday) before it gets merged into `devel` which will be `2.4` at that point

**Comment by joshmoore** [comment]
ready_for_review

**Comment by joshmoore** [comment]
@nerzhul : you ok to give a green review based on https://github.com/ansible/ansible/pull/20627#issuecomment-284366643 ? (Think it's a GitHub bug since the commits that you commented on have now been force pushed away)

**Review by nerzhul — APPROVED** [review]

**Comment by nerzhul** [comment]
It's okay for me but your change cannot be merged for 2.3, maybe it should be changed to 2.4 @gundalow ?

File: lib/ansible/modules/database/postgresql/postgresql_db.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_db.py]: `"2.4"`

**Comment by joshmoore** [comment]
ready_for_review

**Comment by mattclay** [comment]
Closing and re-opening to trigger CI.

**Comment by joshmoore** [comment]
ready_for_review

**Review by gundalow — APPROVED** [review]

**Comment by nerzhul** [comment]
shipit

**Comment by pchauncey** [comment]
Might just be me/user error, but I'm getting this on both ansible 2.3 and ansible 2.4: 

```
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "db": "foo", 
            "encoding": "", 
            "lc_collate": "", 
            "lc_ctype": "", 
            "login_host": "127.0.0.1", 
            "login_password": "", 
            "login_unix_socket": "", 
            "login_user": "postgres", 
            "name": "foo", 
            "owner": "", 
            "port": 5432, 
            "ssl_mode": "prefer", 
            "ssl_rootcert": null, 
            "state": "dump", 
            "target": "/tmp/foo.sql.gz", 
            "template": ""
        }
    }, 
    "msg": "db_dump() got an unexpected keyword argument 'sslmode'"
}


ansible 2.4.0 (testing_PR20627 b4bffa6d8a) last updated 2017/05/08 10:23:49 (GMT -500)
postgresql-9.5.6-1.fc25.x86_64
```
I've tried with ssl_mode=disable with the same results

**Comment by joshmoore** [comment]
I've reproduced in an integration test:

```
TASK [postgresql : Dump the db] ************************************************
fatal: [testhost]: FAILED! => {"changed": false, "failed": true, "msg": "db_dump() got an unexpected keyword argument 'sslmode'"}
	to retry, use: --limit @/root/ansible/test/integration/postgresql-y_LlPW.retry

```

I'll push a fix shortly.

**Comment by joshmoore** [comment]
ready_for_review

Note: there are a couple of changes that I could see adding on top of this even for 2.4 but would prefer to get this in first since it's been something of a struggle with the repo migration and both the 2.2 and 2.3 releases. Features to follow include:

 * `-Fd` and `-j $CPUS` to allow saving a database in parallel
 * ~~support for `login_password`~~ (see https://github.com/ansible/ansible/pull/20627/commits/0939ccab45a96422a5f328110d253ebd4900f22a)

**Comment by pchauncey** [comment]
Now it appears that the dump and restore actions aren't honoring login_user:

The "present" action works:
```
ansible localhost -m postgresql_db -a "name=foo login_host=127.0.0.1 login_user=postgres state=present" 
localhost | SUCCESS => {
    "changed": false, 
    "db": "foo"
}
```

But the "dump" action doesn't, and it's trying to use my UID instead of postgres:
```
ansible localhost -m postgresql_db -a "name=foo target=/tmp/foo.sql login_host=127.0.0.1 login_user=postgres state=dump" -vvv
localhost | FAILED! => {
    "changed": false, 
    "cmd": "/usr/bin/psql --dbname=foo < /tmp/foo.sql", 
    "failed": true, 
    "invocation": {
        "module_args": {
            "db": "foo", 
            "encoding": "", 
            "lc_collate": "", 
            "lc_ctype": "", 
            "login_host": "127.0.0.1", 
            "login_password": "", 
            "login_unix_socket": "", 
            "login_user": "postgres", 
            "name": "foo", 
            "owner": "", 
            "port": 5432, 
            "ssl_mode": "prefer", 
            "ssl_rootcert": null, 
            "state": "restore", 
            "target": "/tmp/foo.sql", 
            "template": ""
        }
    }, 
    "msg": "psql: FATAL:  role \"patrick\" does not exist\n", 
    "rc": 2, 
    "stdout": "", 
    "stdout_lines": []
}

```

**Comment by joshmoore** [comment]
Thanks for the testing @pchauncey. Good you give https://github.com/ansible/ansible/pull/20627/commits/0939ccab45a96422a5f328110d253ebd4900f22a a try?

**Comment by pchauncey** [comment]
Hi @joshmoore, thanks for the quick updates, it is using the postgres user now, but:  Still getting an error here from the same ansible command.  I tried a playbook as well, but both result in the error below:

```
$ ansible localhost -m postgresql_db -a "name=foo target=/tmp/foo.sql login_host=127.0.0.1 login_user=postgres state=dump" -vvv
localhost | FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "db": "foo", 
            "encoding": "", 
            "lc_collate": "", 
            "lc_ctype": "", 
            "login_host": "127.0.0.1", 
            "login_password": "", 
            "login_unix_socket": "", 
            "login_user": "postgres", 
            "name": "foo", 
            "owner": "", 
            "port": 5432, 
            "ssl_mode": "prefer", 
            "ssl_rootcert": null, 
            "state": "dump", 
            "target": "/tmp/foo.sql", 
            "template": ""
        }
    }, 
    "msg": "Database query failed: putenv() argument 2 must be string, not None"
}
$ ansible --version
ansible 2.4.0 (pr20627 f4c49664f0) last updated 2017/05/10 08:17:28 (GMT -500)
```

**Comment by joshmoore** [comment]
Hi @pchauncey. What this really needs is a number of tests per each combination of user/password, etc. In lieu of that, thanks again for hammering on it!

**Comment by pchauncey** [comment]
Here are a few of quick fixes:

1. In db_dump() host wasn't being passed to pg_dump if db name was supplied (elif should be if)
2. Quotes around host and port were breaking things for pg_dump if specified
3. In db_restore() cmd was only getting the first flag passed (flag[0])

How do these look to you?

```diff
--- a/lib/ansible/modules/database/postgresql/postgresql_db.py
+++ b/lib/ansible/modules/database/postgresql/postgresql_db.py
@@ -241,12 +241,12 @@ def db_dump(module, target,
     flags = ''
     if db:
         flags += ' {0}'.format(pipes.quote(db))
-    elif host:
-        flags += ' --host={0}'.format(pipes.quote(host))
-        if port:
-            flags += ' --port={0}'.format(pipes.quote(port))
+    if host:
+        flags += ' --host={0}'.format(host)
+    if port:
+        flags += ' --port={0}'.format(port)
     if user:
-        flags += ' --username={0}'.format(pipes.quote(user))
+        flags += ' --username={0}'.format(user)
 
     cmd = module.get_bin_path('pg_dump', True)
     comp_prog_path = None
@@ -311,7 +311,8 @@ def db_restore(module, target,
     elif os.path.splitext(target)[-1] == '.xz':
         comp_prog_path = module.get_bin_path('xzcat', True)
 
-    cmd += flags[0]
+    for flag in flags:
+        cmd += flag
 
     if comp_prog_path:
         p1 = subprocess.Popen([comp_prog_path, target], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

**Comment by joshmoore** [comment]
Thanks for the fixes, @pchauncey. They certainly pointed to further issues (both in the code and in the tests). I've applied the patch with your info and have pushed a few commits on top of it. Looking forward to more feedback.

**Review by pchauncey — APPROVED** [review]
shipit

**Comment by pchauncey** [comment]
Looks great.  Thanks for all of the additional changes!

Tested on ansible 2.4.0 with PostgreSQL 9.5.6

**Comment by joshmoore** [comment]
Gladly! :) Eventually I could use some info on the minimum supported postgres version and similar or perhaps just on how to scope a feature to a given version in order to introduce the `-j` option. But definitely looking forward to calling this PR done first.

**Comment by joshmoore** [comment]
Anything else needed from me here?

**Comment by pchauncey** [comment]
bot_status

**Comment by joshmoore** [comment]
Should I close & re-open or will that put us back at square one?

**Comment by gundalow** [comment]
I've clicked "rebuild" on shippable, which should remove the `stale_ci` label.

If people are happy after than I think we can merge.

**Comment by joshmoore** [comment]
bot_status

**Comment by joshmoore** [comment]
Sorry, @gundalow, but how do I go about making sure everyone is happy?

**Comment by gundalow** [comment]
Closing and reopening to trigger CI

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
@matburt @pchaunceyn are you happy with this?

**Comment by gundalow** [comment]
I'll merge this next week if I don't hear anything otherwise.

**Comment by joshmoore** [comment]
Friendly reminder.

**Comment by pchauncey** [comment]
Looks good to me

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Thanks for the PR and reviews.
As this is new functionality this has been merged into `devel` and will be released in `2.4.0`

**Comment by autotune** [comment]
@joshmoore additional thanks from me as well for helping drive this through to completion.

**Pull request merged**
